azurerm/internal/services/eventhub/eventhub_cluster_resource.go

```diff
@@ -0,0 +1,189 @@
+	client := meta.(*clients.Client).Eventhub.ClusterClient
+	ctx, cancel := timeouts.ForCreateUpdate(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+	log.Printf("[INFO] preparing arguments for Azure ARM EventHub Cluster creation.")
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	cluster := eventhub.Cluster{
+		Location: utils.String(azure.NormalizeLocation(d.Get("location").(string))),
+		Tags:     tags.Expand(d.Get("tags").(map[string]interface{})),
+		Sku:      expandEventHubClusterSkuName(d.Get("sku_name").(string)),
+	}
+
+	future, err := client.Put(ctx, resourceGroup, name, cluster)
+	if err != nil {
+		return err
+	}
+
+	if err := future.WaitForCompletionRef(ctx, client.Client); err != nil {
+		return fmt.Errorf("creating eventhub cluster: %+v", err)
+	}
+
+	read, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil || *read.ID == "" {
+		return fmt.Errorf("cannot read EventHub Cluster %s (resource group %s) ID", name, resourceGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmEventHubClusterRead(d, meta)
+}
+
+func resourceArmEventHubClusterRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Eventhub.ClusterClient
+	ctx, cancel := timeouts.ForRead(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resourceGroup := id.ResourceGroup
+	name := id.Path["clusters"]
+	resp, err := client.Get(ctx, resourceGroup, name)
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("making Read request on Azure EventHub Cluster %q (resource group %q): %+v", name, resourceGroup, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resourceGroup)
+	d.Set("sku_name", flattenEventHubClusterSkuName(resp.Sku))
+	if location := resp.Location; location != nil {
+		d.Set("location", azure.NormalizeLocation(*location))
+	}
+
+	return tags.FlattenAndSet(d, resp.Tags)
+}
+
+func resourceArmEventHubClusterDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*clients.Client).Eventhub.ClusterClient
+	ctx, cancel := timeouts.ForDelete(meta.(*clients.Client).StopContext, d)
+	defer cancel()
+	id, err := azure.ParseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resourceGroup := id.ResourceGroup
+	name := id.Path["clusters"]
+
+	// The EventHub Cluster can't be deleted until four hours after creation so we'll keep retrying until it can be deleted.
+	return resource.Retry(d.Timeout(schema.TimeoutDelete), func() *resource.RetryError {
+		future, err := client.Delete(ctx, resourceGroup, name)
+		if err != nil {
+			if response.WasNotFound(future.Response()) {
+				return nil
+			}
+			if strings.Contains(err.Error(), "Cluster cannot be deleted until four hours after its creation time") {
```

Comment:
Is it worth a 429 check here too given the length of time this operation could be running for?

Reply:
Creating/deleting clusters is actually quite fast. It's just that once created you can't delete it for the first 4 hours. I don't think a case where TF would wait for 4 hours makes sense?

Reply:
Hi @favoretti 
In this case, if the delete operation is requested before the 4 hour minimum lifetime (say as part of a daily build, or acceptance test), the provider is capable of performing the operation without error or intervention (whilst honouring any custom `timeouts`). If the request is sent after that minimum interval this section of code won't trigger and will, as you point out, complete quickly.